consensus/tendermint/messages.go

```diff
@@ -0,0 +1,110 @@
+package tendermint
+
+// Todo: Signature over the messages needs to be handled somewhere. There are 2 options:
+//	1. Add the signature to each message and extend the Validator Set interface to include VerifyMessageSignature
+//	method.
+//	2. The P2P layer signs the message before gossiping to other validators and verifies the signature before passing
+//	the message to the consensus engine.
+//  The benefit of P2P layer handling the verification of the signature is the that the consensus layer can assume
+//  the message is from a validator in the validator set. However, this means that the P2P layer would need to be aware
+//  of the validator set and would need access to the blockchain which may not be a good idea.
+
+type Message[V Hashable[H], H Hash, A Addr] interface {
+	Proposal[V, H, A] | Prevote[H, A] | Precommit[H, A]
+}
+
+type Proposal[V Hashable[H], H Hash, A Addr] struct {
+	Height     uint
+	Round      uint
+	ValidRound int
+	Value      *V
+
+	Sender A
+}
+
+type Prevote[H Hash, A Addr] struct {
+	Vote[H, A]
+}
+
+type Precommit[H Hash, A Addr] struct {
+	Vote[H, A]
+}
```

Comment:
Unless we're adding more fields to this struct, I think this is cheaper:
```suggestion
type Prevote[H Hash, A Addr] Vote[H, A]

type Precommit[H Hash, A Addr] Vote[H, A]
```

Reply:
I agree, I think it is better in sense of readability. I am not sure there would be performance improvements (@infrmtcs do you mean by cheaper that it was going to be more performant?)

Reply:
done